Cancel button for InatImportJob

JoeCohen opened the pull request:
Add a functioning Cancel button to InatImport#show view

- Adds a binary `InatImport#cancel` attribute
- Clicking the Cancel button sets `cancel == true`
- ImportJob checks that attribute deep inside the `parsing?` loop, and even deeper in `ObservationImporter`, terminating the Job if `canceling?`
- Resolves #3080

### Suggested Manual Test

- Create several sacrificial iNat observations
- Read the rest of this test, then import those observations.
- Wait until the `iNat Import Tracker` `Status` changes to `Importing`, then _immediately_ click the Cancel button.
Expected result:
  - The Job terminates, with `Status: Done`
  - It imports fewer than all of your sacrificial iNat observations. (See screenshot below.)
  - `Estimated Remaining Time: 00:00:00`
  - It displays `Ended` time
  
<img width="309" height="286" alt="Screenshot 2025-07-31 at 5 06 33 PM" src="https://github.com/user-attachments/assets/3bad7a78-770e-4088-bdfd-66823cdbdf99" />

- I suggest deleting the sacrificial iNat observations when you're done.

JoeCohen committed:
Add binary `cancel` column to InatImport

JoeCohen committed:
Add InatImport#cancel action setting cancel == true

JoeCohen committed:
Add Cancel button to view, linking to inat_import_cancel_path

JoeCohen committed:
Reset `cancel` when starting an import

JoeCohen committed:
ObservationImporter stops importing if import.canceled?

JoeCohen committed:
Test canceling in JobTest

JoeCohen committed:
Improve ObservationImporter documentation

- Removes comment left over from cut-and-paste
- More precise failure message

JoeCohen committed:
Fix Layout/EmptyLinesAfterModuleInclusion offense

JoeCohen committed:
Revise PageParser documentation

JoeCohen committed:
Document InatObservationImporter

JoeCohen committed:
Simplify `parsing?`, removing unneeded submethod calls

JoeCohen committed:
Move restricted_user_login to preserve reading order

JoeCohen committed:
Delelgate more stuff to inat_import

Improves data/ink ratio

nimmolo commented on app/views/controllers/inat_imports/show.html.erb:
```diff
@@ -31,3 +31,10 @@
+    :inat_import_tracker_cancel.l,
+    inat_import_cancel_path(id: @tracker.inat_import),
+    { class: "mt-3 btn btn-default" }
+  )
```
This will need `data: { turbo_stream: true }`. `class` and `data` do not need to be inside a hash.

nimmolo replied:
Also I'd strongly vote for using the translation string `:CANCEL.l` instead of MO's unfortunate practice of proliferating duplicate translation strings. We are eventually going to have to weed out all the duplicates if we are ever going to use JS front-end translation, because the data size for all these keys and values is unmanageable.

nimmolo commented on config/routes.rb:
```diff
@@ -446,6 +446,8 @@ def route_actions_hash
   get("inat_imports/authorization_response",
       to: "inat_imports#authorization_response",
       as: "inat_import_authorization_response")
+  get("inat_imports/cancel/:id", to: "inat_imports#cancel",
+                                 as: "inat_import_cancel")
```
By Rails convention, routes that are used to change db values should be `post` or `put` routes. We have a few remaining violations of this convention but i did a lot of work getting rid of them.

nimmolo replied:
Making this work is as easy as making this route use `put` instead of `get`, and then using `put_button` instead of `link_to`.

nimmolo commented on app/controllers/inat_imports_controller.rb:
```diff
@@ -175,4 +177,16 @@ def fresh_tracker(inat_import)
+
+  public
+
+  def cancel
+    inat_import = InatImport.find(params[:id])
+    inat_import.update(cancel: true)
+    tracker = InatImportJobTracker.where(inat_import: inat_import).
+              order(:created_at).last
+    # flash_warning(:inat_import_tracker_canceled.l)
+    redirect_to(inat_import_path(inat_import,
+                                 params: { tracker_id: tracker.id }))
+  end
```
I'm not sure we do need a Turbo response here, since we're not updating the tracker page, we're redirecting to the import show page. I'm curious what's not working.

JoeCohen commented on app/classes/inat/observation_importer.rb:
```diff
@@ -1,16 +1,23 @@
 
     def import_page(page)
       page["results"].each do |result|
+        return false if inat_import.canceled?
```
@nimmolo: Success! Thank you so much. I'm glad one of us has more brains than a scarecrow.
I will try making the other changes you suggest:

- `POST` instead of `GET` (i hadn't figured out how to do that, but your hints should get me there)
- `:CANCEL.l`

then get back to you. (There is one issue -- the js controller keeps firing after the Job is canceled via the Cancel button. I'll see if I can figure that out.)

nimmolo replied:
OK. The Stimulus controller won't have gotten the news that the job is canceled, unless your controller response replaces or updates the partial with that data. Your JS controller is not currently capable of polling the DB for that info, and even if it were, it would be inefficient.

nimmolo replied:
You could alter the Stimulus controller to watch the Cancel button directly, and set a data attribute, or otherwise stop polling, as soon as that button is clicked. That would eliminate having to update the partial.

nimmolo replied:
`helpers/link_helpers.rb#put_button`

JoeCohen replied:
> You could alter the Stimulus controller to watch the Cancel button directly, and set a data attribute, or otherwise trigger some mechanism that stops polling, as soon as that button is clicked. That would eliminate having to update the partial.

Could the Stimulus controller app/javascript/controllers/inat-import-job_controller.js watch `InatImport#state` and stop polling when it's `Done`? Else the `iNat Import Tracker` page will likely display incorrect data, e.g., # of obss imported, time remaining, etc.
I thought it was doing that with
```js
  check_if_we_are_done() {
    if (this.currentTarget.dataset.status == "Done" &&
      this.intervalId != null) {
      clearInterval(this.intervalId)
    }
  }
```

nimmolo replied:
Meanwhile - when manual testing, be sure `data-status` is actually getting set to "done" in the partial when you click the cancel button.

JoeCohen committed:
Reload inat_import before checking `canceled?`

Fixes `Cancel` button not working.

JoeCohen committed:
Use app-wide xlation strings

- `:CANCEL.l` instead of `:inat_import_tracker_cancel.l`
- `:STATUS.l` instead of `:inat_import_tracker_status.l`

JoeCohen committed:
Remove stray period

Fixes `ActionView::SyntaxErrorInTemplate` in prior commit

JoeCohen committed:
Call `l` on xlation string, fixing prior commit failure

JoeCohen committed:
replace link_to with put_button

JoeCohen committed:
Change route for cancel action from GET to PUT

JoeCohen committed:
Always show 0 remaining time when Job is Done

Fixes an issue where it showed "Calculating .." because Job was done before computing remaining time.

JoeCohen committed:
Render :show, instead of redirecting there when canceling

Makes Stimulus controller stop when Job is canceled

JoeCohen committed:
Revise test expectation to reflect change in prior commit

JoeCohen commented on app/jobs/inat_import_job.rb:
```diff
@@ -91,31 +107,19 @@ def parsing?(parser)
     return false if page_empty?(parsed_page)
 
-    import_page(parsed_page)
+    observation_importer.import_page(parsed_page)
```
You're hallucinating.

JoeCohen committed:
Delete dead code

JoeCohen commented on app/classes/inat/observation_importer.rb:
```diff
@@ -1,16 +1,23 @@
     end
 
     def import_page(page)
       page["results"].each do |result|
+        return false if inat_import.reload.canceled?
```
The point is that canceled can change during the loop.

JoeCohen commented on app/controllers/inat_imports_controller.rb:
```diff
@@ -175,4 +177,14 @@ def fresh_tracker(inat_import)
+  public
+
+  def cancel
+    @inat_import = InatImport.find(params[:id])
+    @inat_import.update(cancel: true)
+    @tracker = InatImportJobTracker.where(inat_import: @inat_import).
+               order(:created_at).last
+    render(:show)
```
Won't bother with this.

JoeCohen commented on app/views/controllers/inat_imports/show.html.erb:
```diff
@@ -17,17 +17,26 @@
+      { class: "btn btn-default" }
+    )
+  %>
+  <%=
+    put_button(
+      name: :CANCEL.l,
+      path: inat_import_cancel_path(id: @tracker.inat_import),
+      class: "btn btn-default"
+    )
+  %>
```
Won't bother at this stage.

JoeCohen commented:
@nimmolo -- Thanks for the review. I'll deploy it now.
